pkg/ecdsa/retry/retry.go

```diff
@@ -83,3 +83,210 @@ func EvaluateRetryParticipantsForSigning(
+	retryParticipantsCount int,
+) ([]chain.Address, int, bool) {
+	tripletIndexes := make([][3]int, 0, len(operators)*len(operators)*len(operators))
+	for i := 0; i < len(operators)-2; i++ {
+		for j := i + 1; j < len(operators)-1; j++ {
+			for k := j + 1; k < len(operators); k++ {
+				leftOperator := operators[i]
+				middleOperator := operators[j]
+				rightOperator := operators[j]
+				count := len(groupMembers) -
+					int(operatorToSeatCount[leftOperator]) -
+					int(operatorToSeatCount[middleOperator]) -
+					int(operatorToSeatCount[rightOperator])
+
+				if count >= int(retryParticipantsCount) {
+					tripletIndexes = append(tripletIndexes, [3]int{i, j, k})
+				}
+			}
+		}
+	}
+	if index < len(tripletIndexes) {
+		rng.Shuffle(len(tripletIndexes), func(i, j int) {
+			tripletIndexes[i], tripletIndexes[j] = tripletIndexes[j], tripletIndexes[i]
+		})
+		triplet := tripletIndexes[index]
+		leftOperator := operators[triplet[0]]
+		middleOperator := operators[triplet[1]]
+		rightOperator := operators[triplet[2]]
+		usedOperators := make([]chain.Address, 0, len(groupMembers))
+		for _, operator := range groupMembers {
+			if operator != leftOperator && operator != middleOperator && operator != rightOperator {
+				usedOperators = append(usedOperators, operator)
+			}
+		}
+		return usedOperators, 0, true
+	} else {
+		return nil, len(tripletIndexes), false
+	}
+}
+
+// EvaluateRetryParticipantsForKeyGeneration takes in a slice of `groupMembers`
+// and returns a subslice of those same members of length >=
+// `retryParticipantsCount` randomly according to the provided `seed` and
+// `retryCount`.
+//
+// This function is intended to be called during key generation after a failure
+// *not* due to inactivity. Assuming that some of the `groupMembers` are
+// sending corrupted information, either on purpose or accidentally, we keep
+// trying to find a subset of `groupMembers` that is as large as possible by
+// first excluding single operators, then pairs of operators, then triplets of
+// operators.
+//
+// The `seed` param needs to vary on a per-message basis but must be the same
+// seed between all operators for each invocation. This can be the hash of the
+// message since cryptographically secure randomness isn't important.
+//
+// The `retryCount` denotes the number of the given retry, so that should be
+// incremented after each attempt while the `seed` stays consistent on a
+// per-message basis.
+func EvaluateRetryParticipantsForKeyGeneration(
+	groupMembers []chain.Address,
+	seed int64,
+	retryCount uint,
+	retryParticipantsCount uint,
+) ([]chain.Address, error) {
+	originalRetryCount := retryCount
+	if int(retryParticipantsCount) > len(groupMembers) {
+		return nil, fmt.Errorf(
+			"asked for too many seats. [%d] seats were requested, "+
+				"but there are only [%d] available.",
+			retryParticipantsCount,
+			len(groupMembers),
+		)
+	}
+	operatorToSeatCount := calculateSeatCount(groupMembers)
+	source := rand.NewSource(seed)
```

Comment:
We needed `seed+retryCount` in `EvaluateRetryParticipantsForSigning` because the way we were getting different operator subsets for each seed+retry combo is to shuffle the operators based `seed+retryCount` and then draw operators until we have enough. Changing either seed or retry count changes the shuffle.

Here, we're shuffling the singes/pairs/triplets of operators based on the `seed`, and then excluding one of them based on the retryCount. If there are 85 viable singles, 500 viable pairs, and 3000 viable triplets, and we're handed a `retryCount` of 3200, then we're going to generate all of the viable triplets, shuffle it according to `seed`, and then select the `3200 - 85 - 500 = 2_615` index of the viable triplet list to *exclude* from our operator set.

Each seed would give us a different shuffling, and each `retryCount` would exclude different operators

Reply:
[59d7e92](https://github.com/keep-network/keep-core/pull/3119/commits/59d7e92181447647d096603b1aab39aedfa9c2e8
)

Reply:
Makes sense! Thanks!